index.js

```diff
@@ -202,6 +202,11 @@ NYC.prototype.instrumentAllFiles = function (input, output, cb) {
     var ext
     var transform
     var inFile = path.resolve(inputDir, filename)
+
+    if (!_this.exclude.shouldInstrument(filename)) {
+      return
+    }
+
     var code = fs.readFileSync(inFile, 'utf-8')
 
     for (ext in _this.transforms) {
```

Comment:
This results in no output for files that are not instrumented.  But if I have:
```
src/index.js
src/no-coverage.js
```
With `src/no-coverage.js` excluded, I think I still need both `dest/index.js` and `dest/no-coverage.js` to be created, just `dest/no-coverage.js` should be unmodified.  Same if I run `nyc instrument src/no-coverage.js` - that should print the original source.

So I'm thinking:
```
if (_this.exclude.shouldInstrument) {
  for (ext in ...) { ... }
}
```

I suspect this will change some of the tests, you will have to look for `cov_` or not in the output files for example instead of checking if the file exists/doesn't.

Reply:
If I understand correctly, you're thinking your point is that `dest/` should have a fully functional codebase, just some files will be instrumented and some won't. I agree, it would be a bit shocking to be missing files, it makes it impossible to import the library.

I believe we check for the `cov__` global in a few other tests, I'd look for an existing test using this approach.

Reply:
Yes, though this idea might require bypassing `walkAllFiles` entirely, just using:
```js
glob.sync('**/*', { cwd: input, nodir: true }).forEach(visitor)
```

Specifically `walkAllFiles` uses the extension and exclude filters which we don't really want if we're attempting to copy everything from src to dest.  This could be problematic though when using `npx nyc instrument . dest` as that would include `node_modules`.  Maybe `nyc instrument` needs a `--full-copy` option?  If enabled we directly use `glob.sync` to capture all files, if not then we use `testExclude.globSync`.  I'm leaning towards leaving that option disabled by default.

Possibly future enhancement: when `nyc instrument` copies files which have `chmod +x` enabled should we do the same for the copied file?

Reply:
Let's create an issue for keeping the perms the same; I think it would potentially even be worth splitting this into its own library -- it sounds like we want something like `cpr` that:

* allows us to apply our own `include`/`exclude` rules.
* maintains file permissions.

Definitely beyond the scope of this pull however; although I think maintaining the existing behavior of copying over the whole directory is smart.

Reply:
I've pushed a new commit that addresses some of these concerns, namely copying all existing files in `input` to output before instrumentation and replacing output files in place.  To do this I've brought in `fs-extra` as it seems like a popular and maintained module, but I'm not particularly attached to it so if it needs to change let me know of your preferred replacement.